tests/drivers/rtio/rtio_api/src/test_rtio.c

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright (c) 2019 Thomas Burdick <thomas.burdick@gmail.com>
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <ztest.h>
+#include <drivers/rtio.h>
+
+RTIO_MEMPOOL_ALLOCATOR_DEFINE(block_pool_alloc, 16, 1024, 1);
+
+void test_rtio_mempool_allocator(void)
+{
+	struct rtio_block *block;
+	struct rtio_block *block2;
+
+	u32_t size = 1024;
+	int res = rtio_block_alloc(&block_pool_alloc, &block, size, K_NO_WAIT);
+
+	zassert_equal(res, 0, "Allocation should not fail");
+	zassert_equal(block->size, size, "Size should be 1024");
+	zassert_equal(block->len, 0, "Len should be zero");
+	zassert_not_null(block->_data, "Data should not be NULL");
+	res = rtio_block_alloc(&block_pool_alloc, &block2, size, K_NO_WAIT);
+
+	zassert_equal(res, -ENOMEM, "Allocation should fail with no memory");
+
+	rtio_block_free(&block_pool_alloc, block);
+
+	res = rtio_block_alloc(&block_pool_alloc, &block2, size, K_NO_WAIT);
+
+	zassert_equal(res, 0, "Allocation should not fail");
+}
+
+RTIO_MEMPOOL_ALLOCATOR_DEFINE(pipe_pool, 16, 1024, 1);
+
+#define STACK_SIZE 512
+K_THREAD_STACK_DEFINE(source_stack, STACK_SIZE);
+static struct k_thread source_thread;
+K_THREAD_STACK_DEFINE(sink_stack, STACK_SIZE);
+static struct k_thread sink_thread;
+
+K_FIFO_DEFINE(block_fifo);
+
+#define EXPECTED_VAL 5
+
+void source_entry(void *fifov, void *p2, void *p3)
+{
+	struct k_fifo *fifo = (struct k_fifo *)fifov;
+	struct rtio_block *block;
+	int res = rtio_block_alloc(&pipe_pool, &block, sizeof(u32_t),
+				   K_NO_WAIT);
+	zassert_equal(res, 0, "Allocation should not fail");
+	rtio_block_begin(block);
+	rtio_block_push_u32_t(block, (u32_t)EXPECTED_VAL);
+	rtio_block_end(block);
+	k_fifo_put(fifo, block);
```

Comment:
From:  https://docs.zephyrproject.org/latest/reference/kernel/data_passing/fifos.html

> FIFO data items must be aligned on a word boundary, as **the kernel reserves the first word of an item for use as a pointer to the next data item in the queue**. 

So the k_fifo_put() will destroy begin_tstamp. Is this intended?

Reply:
100% not intended. Does this mean I need a void* spacer at the beginning of each rtio_block? Is there a better way of doing that?

Reply:
Ok yes, it does, fixed. Not very happy that I need an additional member in rtio block, might reconsider the usage of a fifo over a ringbuffer